services/github/src/api.js

```diff
@@ -474,64 +474,64 @@ builder.declare({
-          checksSoftStates.push(check.conclusion);
-          continue;
-        }
+    // List of conclusions: https://docs.github.com/en/rest/reference/checks#check-runs
+    const SOFT_STATES = ['pending', 'timed_out', 'cancelled', 'skipped', 'stale', 'timed_out', 'action_required', 'neutral'];
+    const checksSoftStates = [];
 
-        // If the current state is success or pending and a check was successful, mark state as success.
-        // This might get reset after the loop if we find any other pending check.
-        if (state !== 'failure' && check.conclusion === 'success') {
-          state = 'success';
-        }
+    for (const check of checks) {
+      // If any check failed, mark the whole status as failed
+      if (check.conclusion === 'failure') {
+        state = 'failure';
+        break;
       }
 
-      if (state !== 'failure' && checksSoftStates.length > 0) {
-        state = checksSoftStates[0];
+      // If there's any check pending, set the state to pending after looping unless there was a failure
+      if (SOFT_STATES.includes(check.conclusion)) {
+        checksSoftStates.push(check.conclusion);
+        continue;
       }
 
-      if (state) {
-        // If we got a status, send a corresponding image.
-        fileConfig.headers['X-Taskcluster-Status'] = state;
-        return res.sendFile(state + '.svg', fileConfig);
-      } else {
-        // otherwise, it's a commit without a TC status, which probably means a new repo
-        fileConfig.headers['X-Taskcluster-Status'] = 'newrepo';
-        return res.sendFile('newrepo.svg', fileConfig);
-      }
-    } catch (e) {
-      if (e.code < 500) {
-        fileConfig.headers['X-Taskcluster-Status'] = 'error';
-        return res.sendFile('error.svg', fileConfig);
+      // If the current state is success or pending and a check was successful, mark state as success.
+      // This might get reset after the loop if we find any other pending check.
+      if (state !== 'failure' && check.conclusion === 'success') {
+        state = 'success';
       }
-      throw e;
     }
-  } else {
-    fileConfig.headers['X-Taskcluster-Status'] = 'nogithub';
-    return res.sendFile('newrepo.svg', fileConfig);
+
+    if (state !== 'failure' && checksSoftStates.length > 0) {
+      state = checksSoftStates[0];
+    }
+
+    if (state) {
+      // If we got a status, send a corresponding image.
+      fileConfig.headers['X-Taskcluster-Status'] = state;
+      return res.sendFile(state + '.svg', fileConfig);
+    } else {
+      // otherwise, it's a commit without a TC status, which probably means a new repo
+      fileConfig.headers['X-Taskcluster-Status'] = 'newrepo';
+      return res.sendFile('newrepo.svg', fileConfig);
+    }
+  } catch (e) {
+    if (e.code < 500) {
+      fileConfig.headers['X-Taskcluster-Status'] = 'error';
+      return res.sendFile('error.svg', fileConfig);
+    }
+    res.reportError('InternalError', e.message, {});
```

Comment:
I wonder if we should send the original exception message back to client, or only log it

Reply:
It would be nice if a client could figure out what went wrong without having to acess logs. Would be nice if it could report when something wasn't found.

Reply:
alright, makes sense. I hope there will be nothing sensitive in those exceptions. But should be safe, as those would be most likely coming from github itself